File: ios/Sources/NoiseTestHarness/NoiseVectorFixtureRepository.swift
import Foundation

public actor NoiseVectorFixtureRepository {
    private let loader: NoiseVectorFixtureLoader
    private var cachedCatalog: NoiseVectorFixtureCatalog?

    public init(loader: NoiseVectorFixtureLoader = .init()) {
        self.loader = loader
    }

    public func catalog() throws -> NoiseVectorFixtureCatalog {
        if let cachedCatalog {
            return cachedCatalog
        }

        let catalog = try NoiseVectorFixtureCatalog(fixtures: loader.loadFixtures())
        cachedCatalog = catalog
        return catalog
    }

    public func fixtures() throws -> [NoiseVectorFixture] {
        try catalog().all()
    }

    public func fixture(vectorID: String) throws -> NoiseVectorFixture {
        try catalog().fixture(vectorID: vectorID)
    }

    public func filter(
        pattern: NoiseVectorPattern? = nil,
        diffieHellman: NoiseVectorDiffieHellman? = nil,
        cipher: NoiseVectorCipher? = nil,
        hash: NoiseVectorHash? = nil
    ) throws -> [NoiseVectorFixture] {
        try catalog().filter(
            pattern: pattern,
            diffieHellman: diffieHellman,
            cipher: cipher,
            hash: hash
        )
    }
}

public struct NoiseVectorFixtureCatalog: Sendable, Equatable {
    private let allFixtures: [NoiseVectorFixture]
    private let fixturesByID: [String: NoiseVectorFixture]

    public init(fixtures: [NoiseVectorFixture]) throws {
        let indexed = Dictionary(grouping: fixtures, by: \ .vectorID)
        if let duplicate = indexed.first(where: { $0.value.count > 1 }) {
            throw NoiseTestHarnessError.invalidFixture(
                "Fixture corpus contains duplicate vector_id values: \(duplicate.key)."
            )
        }

        self.allFixtures = fixtures
        self.fixturesByID = indexed.mapValues { $0[0] }
    }

    public func all() -> [NoiseVectorFixture] {
        allFixtures
    }

    public func fixture(vectorID: String) throws -> NoiseVectorFixture {
        guard let fixture = fixturesByID[vectorID] else {
            throw NoiseTestHarnessError.fixtureFileNotFound(vectorID)
        }
        return fixture
    }

    public func filter(
        pattern: NoiseVectorPattern? = nil,
        diffieHellman: NoiseVectorDiffieHellman? = nil,
        cipher: NoiseVectorCipher? = nil,
        hash: NoiseVectorHash? = nil
    ) -> [NoiseVectorFixture] {
        allFixtures.filter { fixture in
            (pattern == nil || fixture.protocolInfo.pattern == pattern) &&
                (diffieHellman == nil || fixture.protocolInfo.suite.dh == diffieHellman) &&
                (cipher == nil || fixture.protocolInfo.suite.cipher == cipher) &&
                (hash == nil || fixture.protocolInfo.suite.hash == hash)
        }
    }
}

File: android/noise-testing/src/main/kotlin/noise/protocol/testing/NoiseVectorFixtureRepository.kt
package noise.protocol.testing

import noise.protocol.core.HandshakePattern
import java.nio.file.Path

class NoiseVectorFixtureRepository(
    private val fixturesDirectory: Path,
    private val loader: NoiseVectorFixtureLoader = NoiseVectorFixtureLoader()
) {
    @Volatile
    private var cachedCatalog: NoiseVectorFixtureCatalog? = null

    fun catalog(): NoiseVectorFixtureCatalog {
        val existing = cachedCatalog
        if (existing != null) {
            return existing
        }

        return synchronized(this) {
            cachedCatalog ?: loadCatalog().also { cachedCatalog = it }
        }
    }

    fun all(): List<NoiseVectorFixture> = catalog().all()

    fun findById(vectorId: String): NoiseVectorFixture? = catalog().findById(vectorId)

    fun requireById(vectorId: String): NoiseVectorFixture = catalog().requireById(vectorId)

    fun filter(
        pattern: HandshakePattern? = null,
        dh: VectorDhAlgorithm? = null,
        cipher: VectorCipherAlgorithm? = null,
        hash: VectorHashAlgorithm? = null
    ): List<NoiseVectorFixture> {
        return catalog().filter(
            pattern = pattern,
            dh = dh,
            cipher = cipher,
            hash = hash
        )
    }

    private fun loadCatalog(): NoiseVectorFixtureCatalog {
        return NoiseVectorFixtureCatalog(loader.loadAll(fixturesDirectory))
    }
}

class NoiseVectorFixtureCatalog internal constructor(
    fixtures: List<NoiseVectorFixture>
) {
    private val allFixtures: List<NoiseVectorFixture> = fixtures.toList()
    private val fixturesById: Map<String, NoiseVectorFixture> = allFixtures.associateBy { it.vectorId }

    init {
        require(fixturesById.size == allFixtures.size) {
            "Fixture corpus contains duplicate vector_id values."
        }
    }

    fun all(): List<NoiseVectorFixture> = allFixtures

    fun findById(vectorId: String): NoiseVectorFixture? = fixturesById[vectorId]

    fun requireById(vectorId: String): NoiseVectorFixture {
        return findById(vectorId)
            ?: error("Fixture '$vectorId' does not exist in the loaded corpus.")
    }

    fun filter(
        pattern: HandshakePattern? = null,
        dh: VectorDhAlgorithm? = null,
        cipher: VectorCipherAlgorithm? = null,
        hash: VectorHashAlgorithm? = null
    ): List<NoiseVectorFixture> {
        return allFixtures.filter { fixture ->
            (pattern == null || fixture.protocol.pattern == pattern) &&
                (dh == null || fixture.protocol.suite.dh == dh) &&
                (cipher == null || fixture.protocol.suite.cipher == cipher) &&
                (hash == null || fixture.protocol.suite.hash == hash)
        }
    }
}